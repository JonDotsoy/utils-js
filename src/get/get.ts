/**
 * Recursively retrieves a property from an object by following the specified path.
 *
 * @param {unknown} obj - The object to retrieve the property from.
 * @param {...PropertyKey[]} paths - The path(s) of properties to follow. If no paths are provided, returns the entire object.
 * @returns {unknown} The retrieved value or undefined if the path is invalid.
 */
export const get = (obj: unknown, ...paths: PropertyKey[]): unknown => {
  if (paths.length === 0) return obj;
  const isObj = typeof obj === "object" && obj !== null;
  if (!isObj) return undefined;
  const [path, ...nextPaths] = paths;
  return get(Reflect.get(obj, path), ...nextPaths);
};

type types = "string" | "number" | "boolean" | "function" | "bigint" | "symbol";

const createValidatorPrimitiveType =
  <T>(type: types[number]) =>
  (obj: unknown, ...paths: PropertyKey[]): undefined | T => {
    const value = get(obj, ...paths);
    if (typeof value !== type) return undefined;
    return value as T;
  };

const createValidatorCustomType =
  <T>(test: (value: unknown) => boolean) =>
  (obj: unknown, ...paths: PropertyKey[]): undefined | T => {
    const value = get(obj, ...paths);
    if (!test(value)) return undefined;
    return value as T;
  };

/** Validates that a value is an string */
get.string = createValidatorPrimitiveType<string>("string");
/** Validates that a value is an number */
get.number = createValidatorPrimitiveType<number>("number");
/** Validates that a value is an boolean */
get.boolean = createValidatorPrimitiveType<boolean>("boolean");
/** Validates that a value is an function */
get.function = createValidatorPrimitiveType<Function>("function");
/** Validates that a value is an BigInt */
get.bigint = createValidatorPrimitiveType<bigint>("bigint");
/** Validates that a value is an symbol */
get.symbol = createValidatorPrimitiveType<symbol>("symbol");
/** Validates that a value is an Array */
get.array = createValidatorCustomType<Array<unknown>>((value) =>
  Array.isArray(value),
);

/** Validates that a value is an Date */
get.date = createValidatorCustomType<Date>((value) => {
  if (value instanceof Date) return true;
  if (typeof value === "string") return !isNaN(Date.parse(value));
  return false;
});

/** Validates that a value is an object */
get.record = createValidatorCustomType<Record<any, any>>(
  (value) => typeof value === "object" && value !== null,
);
