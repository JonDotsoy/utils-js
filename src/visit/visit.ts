import { get } from "../get/get.js";

/**
 * A type alias for a function that takes an unknown node as input and returns a boolean.
 */
type Test<A> = (node: A) => boolean;

/**
 * A generator function that recursively visits nodes in an object, yielding each node that passes the provided test.
 *
 * @param {T} node The starting node to visit.
 * @param {Test} [test] An optional function that takes a node as input and returns a boolean. If true, the node will be yielded.
 * @returns A generator of nodes that pass the test.
 */
export function* visit<T, A = T, R = T>(node: T, test?: Test<A>): Generator<R> {
  if (test?.(node as any) ?? true) yield node as unknown as R;
  const obj = get.record(node);
  for (const key in obj) {
    yield* visit(obj[key], test);
  }
}
