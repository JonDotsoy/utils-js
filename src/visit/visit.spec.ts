import { expect, test, mock } from "bun:test";
import { visit } from "./visit.js";
import { get } from "../get/get.js";

test("should return once node", () => {
  const fn = mock();
  const obj = {
    a: {
      b: [{ c: [{ visitable: true, d: 3 }] }],
    },
  };

  for (const elm of visit(obj, (e) => get.boolean(e, "visitable") ?? false)) {
    fn(elm);
  }

  expect(fn).toHaveBeenCalledTimes(1);
});

test("should return every nodes available", () => {
  const v = visit([1, 2, 3]);
  expect([...v]).toMatchSnapshot();
});

test("should return only nodes of type span", () => {
  const v = visit(
    {
      children: [
        { type: "span", value: "foo" },
        { type: "block", children: [{ type: "span", value: "var" }] },
      ],
    },
    (node: any) =>
      typeof node === "object" && node !== null && node.type === "span",
  );
  expect([...v]).toMatchSnapshot();
});

test("should return the parent element", () => {
  const object = {
    a: {
      b: [
        { c: { name: 1, toVisit: true } },
        [{}, { name: 4, toVisit: true }],
        { [Symbol("sd")]: { name: 3, toVisit: true } },
      ],
    },
    e: {
      name: 2,
      toVisit: true,
    },
  };

  expect(
    Array.from(
      visit(object, (object) => get.boolean(object, "toVisit") ?? false),
      (child) => ({
        field: visit.getFieldName(child),
        parent: visit.getParent(child),
      }),
    ),
  ).toMatchSnapshot();
});

test.only("should visit recursive node once time", () => {
  const nodeReferenced: Record<string, any> = {
    name: "foo",
  };

  nodeReferenced.recursive = nodeReferenced;

  const rootNode = {
    a: { b: nodeReferenced },
    b: [{ c: { d: nodeReferenced } }],
  };

  Array.from(visit(rootNode));
});
