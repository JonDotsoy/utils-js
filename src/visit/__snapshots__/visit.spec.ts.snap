// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should return only nodes of type span 1`] = `
[
  {
    "type": "span",
    "value": "foo",
  },
  {
    "type": "span",
    "value": "var",
  },
]
`;

exports[`should return every nodes available 1`] = `
[
  [
    1,
    2,
    3,
  ],
  1,
  2,
  3,
]
`;
